src/core/geometry/qgsgeos.cpp

```diff
@@ -1475,11 +1462,23 @@ GEOSGeometry* QgsGeos::createGeosPoint( const QgsAbstractGeometryV2* point, int
   try
   {
     GEOSCoordSequence* coordSeq = GEOSCoordSeq_create_r( geosinit.ctxt, 1, coordDims );
-    GEOSCoordSeq_setX_r( geosinit.ctxt, coordSeq, 0, pt->x() );
-    GEOSCoordSeq_setY_r( geosinit.ctxt, coordSeq, 0, pt->y() );
-    if ( pt->is3D() )
+    if ( precision > 0. )
+    {
+      GEOSCoordSeq_setX_r( geosinit.ctxt, coordSeq, 0, std::round( pt->x() / precision ) * precision );
```

Comment:
Same here. May I suggest to use a function for snapping an ordinate value to a precision grid ? 
So you have a single place to test/fix...